File: Python/day10/AOC.py
def look_and_say(input):
    count = 0
    look_at = input[0]
    output = ""
    for char in input:
        if look_at == char:
            count += 1
        else:
            output += str(count) + look_at
            look_at = char
            count = 1
    output += str(count) + look_at
    return output

def part1(input):
    for _ in range(40):
        input = look_and_say(input)
    return len(input)

def part2(input):
    for _ in range(50):
        input = look_and_say(input)
    return len(input)

def main():
    input = open("input.txt", "r").read().splitlines()
    part1_sum = part1(input[0])
    part2_sum = part2(input[0])
    print(f"Part 1: {part1_sum}")
    print(f"Part 2: {part2_sum}")

if __name__ == "__main__":
    main()

File: Go/day10/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func lookAndSay(s string) string {
	var result strings.Builder
	for i := 0; i < len(s); i++ {
		count := 1
		for i+1 < len(s) && s[i] == s[i+1] {
			i++
			count++
		}
		result.WriteString(fmt.Sprintf("%d%c", count, s[i]))
	}
	return result.String()
}

func part1(input []string) int {
	init := input[0]
	for i := 0; i < 40; i++ {
		init = lookAndSay(init)
	}
	return len(init)
}

func part2(input []string) int {
	init := input[0]
	for i := 0; i < 50; i++ {
		init = lookAndSay(init)
	}
	return len(init)
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
